### PR TITLE
on a deref, check that we are not using a mutable ref with a frozen tag

### PR DESCRIPTION
**Pull request by RalfJung**

### CODE REVIEW

**Comment by RalfJung** [comment]
With this and the WIP barriers, there are no more ignored testcases :)

File: src/stacked_borrows.rs [hunk 1]
```diff
@@ -151,6 +151,12 @@ impl<'tcx> Stack {
     /// Returns the index of the item we matched, `None` if it was the frozen one.
     /// `kind` indicates which kind of reference is being dereferenced.
     fn deref(&self, bor: Borrow, kind: RefKind) -> Result<Option<usize>, String> {
+        // Exclude unique ref and frozen tag.
+        match (kind, bor) {
+            (RefKind::Unique, Borrow::Shr(Some(_))) =>
+                return Err(format!("Encountered mutable reference with frozen tag")),
```
Comment (oli-obk) [inline src/stacked_borrows.rs]: The previous error reported some info on the pointer. Might be beneficial here, too.

File: src/stacked_borrows.rs [hunk 2]
```diff
@@ -151,6 +151,12 @@ impl<'tcx> Stack {
     /// Returns the index of the item we matched, `None` if it was the frozen one.
     /// `kind` indicates which kind of reference is being dereferenced.
     fn deref(&self, bor: Borrow, kind: RefKind) -> Result<Option<usize>, String> {
+        // Exclude unique ref and frozen tag.
+        match (kind, bor) {
+            (RefKind::Unique, Borrow::Shr(Some(_))) =>
```
Comment (oli-obk) [inline src/stacked_borrows.rs]: nit: match with just one arm can be an `if let`

**Comment by RalfJung** [comment]
Fixed nits.

**Pull request merged**